modules/fedimint-dummy-client/src/states.rs

```diff
@@ -0,0 +1,82 @@
+use fedimint_client::sm::{DynState, OperationId, State, StateTransition};
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::core::{IntoDynInstance, ModuleInstanceId};
+use fedimint_core::db::ModuleDatabaseTransaction;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::{Amount, TransactionId};
+
+use crate::db::DummyClientFundsKeyV0;
+use crate::{get_funds, DummyClientContext};
+
+/// Tracks a transaction
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum DummyStateMachine {
+    Input(Amount, TransactionId, OperationId),
+    Output(Amount, TransactionId, OperationId),
+    Done,
+}
+
+impl State for DummyStateMachine {
+    type ModuleContext = DummyClientContext;
+    type GlobalContext = DynGlobalClientContext;
+
+    fn transitions(
+        &self,
+        _context: &Self::ModuleContext,
+        global_context: &Self::GlobalContext,
+    ) -> Vec<StateTransition<Self>> {
+        match self.clone() {
+            DummyStateMachine::Input(amount, txid, id) => vec![StateTransition::new(
```

Comment:
Why do we need both of these? They look the same to me

Reply:
It's so the client can track their balance independently of the fed (similar to how the mint module works).

If input fails, we get refunded.  If output fails, we don't.
If input succeeds, nothing happens since we already deducted.  If output succeeds, we add new funds.